isabelle/Core/SemanticsGuarantees.thy

```diff
@@ -5,226 +5,186 @@ imports SemanticsTypes "HOL-Library.Product_Lexorder" Util.MList
 begin
 (*>*)
 
-(* BEGIN Proof of linorder for Party *)
+section "Party linorder"
+
+
+instantiation "Party" :: "ord"
+begin
+
 
 fun less_eq_Party :: "Party \<Rightarrow> Party \<Rightarrow> bool" where
 "less_eq_Party (Address _) (Role _) = True" |
 "less_eq_Party (Role _) (Address _) = False" |
 "less_eq_Party (Address x) (Address y) =  (x \<le> y)" |
 "less_eq_Party (Role x) (Role y) =  (x \<le> y)"
 
-fun less_Party :: "Party \<Rightarrow> Party \<Rightarrow> bool" where
-"less_Party a b = (\<not> (less_eq_Party b a))"
+definition less_Party :: "Party \<Rightarrow> Party \<Rightarrow> bool" where 
+  "less_Party a b \<longleftrightarrow> \<not> ( b \<le> a)"
```

Comment:
If you can choose between having an abstract predicate that is not executable and something that is just a normal computable function, choosing the abstract predicate to save a parenthesis really isn't worth it. If it was a lemma, then I would understand, but it is just a function

Reply:
Not sure what you mean with something abstract that is not executable. This is exported as code as following:
![Screen Shot 2023-04-18 at 18 09 00](https://user-images.githubusercontent.com/2634059/232905644-716eb340-f3d6-4571-a4d2-fafcadc5a121.png)

The "simp problem" mentioned in a previous comment is easily solved here by the line directly below:

```
declare less_Party_def [simp add]
```

which adds the definition to the simp set. I'm using the [principle of least power](https://en.wikipedia.org/wiki/Rule_of_least_power). If you can solve a problem two different ways, and one of them is simpler, the simpler tool should be used. I assume that is the reason why `definition`, `primrec`, `fun` and `function` exist in the first place, and a simple search by text in the Isabelle source code yields

7526 definition 
819 primrec
2241 fun 
1460 function